image.c

```diff
@@ -133,6 +143,141 @@ openImgdisplay()
     return FALSE;
 }
 
+static int
+protocol_test_environment(void)
+{
+    const char * envptr;
+
+    if ((envptr = getenv("TERM"))) {
+	/* The purpose of strncmp(3) where used below is like the glob `prefix*',
+	 * or the regex `^prefix' */
+	if (strcmp(envptr, "xterm-kitty") == 0)
+	    return INLINE_IMG_KITTY;
+	if (strcmp(envptr, "xterm-ghostty") == 0)
+	    return INLINE_IMG_KITTY;
+	if (strncmp(envptr, "mlterm", 6) == 0)
+	    return INLINE_IMG_OSC5379;
+	/* yaft doesn't correctly respond to \e[c, but is sixel-capable
+	 * anyway. Thanks to hackerb9/lsix */
+	if (strncmp(envptr, "yaft", 4) == 0)
+	    return INLINE_IMG_SIXEL;
+    }
+
+    if ((envptr = getenv("KONSOLE_VERSION")) && strcmp(envptr, "220770") >= 0)
+	return INLINE_IMG_KITTY;
+
+    return INLINE_IMG_NONE;
+}
+
+static int
+have_img2sixel(void)
+{
+    pid_t child_pid;
+    int wstatus;
+
+    if (getenv("W3M_IMG2SIXEL"))
+	return TRUE;
+
+    switch (child_pid = fork()) {
+    case -1:
+	return FALSE;
+    case 0:
+	close(STDOUT_FILENO);
+	close(STDERR_FILENO);
+	execlp("img2sixel", "img2sixel", "--version", NULL);
+	/* if exec fails */
+	_exit(-1);
+    default:
+	return
+#ifdef HAVE_WAITPID
+	    waitpid(child_pid, &wstatus, 0) != -1
+#else
+	    wait(&wstatus) != -1
+#endif
+	    && WIFEXITED(wstatus)
+	    && WEXITSTATUS(wstatus) == 0;
+    }
+}
+
+/* NB: in theory, user input could be snarfed up with the read(2); in
+ * practice, only museum pieces have such low latency, so we should get
+ * the device attributes (and only the device attributes) immediately. If
+ * ever this becomes an issue (which it won't), ungetch(3) can be used on
+ * any extraneous input */
+static int
+protocol_test_for_sixel(void)
+{
+    static const char errstr[] = "Can't get terminal attributes";
+
+    /* the string in the following sizeof expression is what I think the
+     * largest possible response could be. See
+     *	https://invisible-island.net/xterm/ctlseqs/ctlseqs.html
+     * for more information, specifically the section on CSI Ps c ("\e[c") */
+    char term_response[sizeof "\033[?65;1;2;3;4;6;8;9;15;16;17;18;21;22;28;29c"] = "";
```

Comment:
FYI, this is by no means the largest possible response. The list of features on the xterm site are just the ones that xterm supports. DEC documented more than 40 of them, and some modern terminals have defined a few of their own.

That said, no one terminal is likely to report all of them, but there will definitely be terminals with longer responses than the one above. For example, this is an example from a real DEC VT520:
```
\033[?65;1;2;6;7;8;9;12;15;18;19;21;23;24;42;44;45;46c
```
A VT525 has at least one more feature than that, so would be even longer.

Reply:
Good point. Plus `read` isn't guaranteed to read the entire response in the first place.
It's better to loop until you see a c.

Reply:
This should also address the problem of snarfing up input on museum pieces, right?

Reply:
No.  Terminal querying is inherently racy, and the only comprehensive solution to that is to feed back any unexpected data before & after the response to an internal buffer (which w3m does not have) for subsequent processing.
In practice, I doubt anybody will run into this - try using images over a high-latency connection to see why :P

Reply:
Well then, still using a loop and a dynamically sized buffer seems the right approach here.